consensus/hotstuff/cruisectl/block_time_controller.go

```diff
@@ -531,6 +531,18 @@ func (ctl *BlockTimeController) EpochEmergencyFallbackTriggered() {
 	ctl.epochFallbacks <- struct{}{}
 }
 
+// EpochExtended handles the epoch extended protocol event.
+func (ctl *BlockTimeController) EpochExtended(extension *flow.EpochExtension) {
+	ctl.curEpochFinalView = extension.FinalView
+	ctl.curEpochTargetEndTime = extension.TargetEndTime
+	ctl.nextEpochFinalView = nil
+}
+
+// EpochRecovered handles the epoch recovered protocol event.
+func (ctl *BlockTimeController) EpochRecovered(nextEpochFinalView *uint64) {
+	ctl.nextEpochFinalView = nextEpochFinalView
+}
```

Comment:
I would recommend to pass `nextEpochFinalView` by value. Internally, the block time controller represents the `nextEpochFinalView` as a pointer, because there might be no next epoch specified atm (e.g. during the staking phase). However, this is the controllers _internal_ way of representing all possible cases.

In contrast, `EpochRecovered` specifically means that the subsequent epoch has now been specified. In this context, a nil value for `nextEpochFinalView` is incompatible with the protocol.